apps/condo/domains/banking/tasks/generateReports.js

```diff
@@ -0,0 +1,270 @@
+// Practically, we need to
+const TASK_PROGRESS_UPDATE_INTERVAL = 10 * 1000 // 10sec
+
+// Rough solution to offload server in case of processing many thousands of records
+const SLEEP_TIMEOUT = conf.WORKER_BATCH_OPERATIONS_SLEEP_TIMEOUT || 200
+
+const DV_SENDER = { dv: 1, sender: { dv: 1, fingerprint: 'generateReports' } }
+
+const logger = getLogger('generateReports')
+
+function sortTransactionsByMonthAndYear (allTransactions) {
+    const YearMonthSet = new Set()
+    allTransactions.map(({ date }) => YearMonthSet.add(dayjs(date).format('YYYY-MM')))
+
+    const setIterator = YearMonthSet.values()
+    const sortedTransactionsByMonthAndYear = []
+    for (let i = 0; i <= YearMonthSet.size - 1; i++ ) {
+        const currentDate = setIterator.next().value
+        const filteredTransactionsByCurrentYearMonth = []
+        for (const transaction of allTransactions) {
+            if (transaction.date.includes(currentDate)) {
+                filteredTransactionsByCurrentYearMonth.push(pickBy(transaction))
+            }
+        }
+        sortedTransactionsByMonthAndYear.push(filteredTransactionsByCurrentYearMonth)
+    }
+    return sortedTransactionsByMonthAndYear
+}
+
+function calculateTransactionsTurnover (sortedTransactions) {
+    let totalIncome = 0, totalOutcome = 0, date
+    const result = []
+    for (let transactions of sortedTransactions) {
+        date = dayjs(transactions[0].date).format('YYYY-MM')
+        for (let transaction of transactions) {
+            transaction.isOutcome ? totalOutcome += Number(transaction.amount) : totalIncome += Number(transaction.amount)
+        }
+        result.push({
+            date,
+            totalIncome: totalIncome.toString(),
+            totalOutcome: totalOutcome.toString(),
+        })
+        totalIncome = 0
+        totalOutcome = 0
+    }
+    return result
+}
+
+async function categoryGroupsBuilder (costItemsDataSortedByDate, { task, lastProgress, context }) {
+    const categoryGroups = []
+    let firstIter = 0, secondIter = 0
+    for (let { date, data } of costItemsDataSortedByDate) {
+        firstIter++
+        let parsedCategoryData = {}
+        for (let costItemId in data) {
+            secondIter++
+            if (Object.keys(parsedCategoryData).includes(data[costItemId].category.id)) {
+                parsedCategoryData[data[costItemId].category.id].costItemGroups.push({
+                    id: costItemId,
+                    name: data[costItemId].name,
+                    sum: data[costItemId].amount,
+                })
+            } else {
+                parsedCategoryData[data[costItemId].category.id] = {
+                    id: data[costItemId].category.id,
+                    name: data[costItemId].category.name,
+                    costItemGroups: [{
+                        id: costItemId,
+                        name: data[costItemId].name,
+                        //isOutcome: data[costItemId].isOutcome,
```

Comment:
@AntonAL this should be uncommented after schema changes

Reply:
Changed schema in https://github.com/open-condo-software/condo/pull/3155